Save posts for later - Expand/collapse animation

malinajirka opened the pull request:
Fixes #7719 

Adds collapse animation when removing an item from saved posts in the saved posts list and expand animation for undo.

To test:
1. Go to Reader - Saved posts list
2. Remove a post from saved posts
3. Notice the item transforms into an Undo item
4. Click on the item
5. Notice the item transforms back to the original item

![expand_collapse_animations](https://user-images.githubusercontent.com/2261188/40181169-bdb3303c-59e8-11e8-8200-e1d6f55d8312.gif)

cc @iamthomasbishop I haven't strictly followed your notes. This is default Android animation and I think it looks pretty good. Implementing our own animation would take extra time and I'm not sure it's worth it. wdyt?

iamthomasbishop commented:
Looking _awesome_, @malinajirka! I have a couple thoughts:

* I noticed that if you only have 1 saved post, then remove it, I don't see the same collapse transition/animation that I do when there are more than 1 posts. Example: https://cloudup.com/c-vBGusVxW3 . I also noticed that the loader/spinner is also appears oddly at the bottom-left of the screen.

* When you collapse a card, if you look closely, you'll see that the contents are actually outside the bounds of the card as they fade out. Can we get the cards' contents to not overflow the bounds of the card while the transition is happening?

* When I mocked this up, I used the old lighter color icons, forgot we had darkened the colors of those during the recent accessibility audit. Can we make this save icon the darker color to match the other action icons?

* I should've noted this earlier, but because we are removing the Share action from the top-level list of actions to make room for Save, I'd like to add Share (or "Share this post") to the ellipsis menu

* We might want to adjust the sizing/spacing of the actions/icons on the list screens in general - that space is a little tight. Not necessary to solve in this issue, but wanted to note it so that we can address this soon - perhaps during the next hack week!

Awesome work! This is looking really great! 👍

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -986,7 +986,9 @@ public void onClick(View v) {
             if (commentStatus == CommentStatus.TRASH) {
-                mBtnModerateIcon.setImageResource(R.drawable.ic_undo_grey_24dp);
+                mBtnModerateIcon.setImageResource(R.drawable.ic_undo_24dp);
+                mBtnModerateIcon.setColorFilter(ContextCompat.getColor(mBtnModerateIcon.getContext(), R.color.grey),
```
;) Good catch! 👍

malinajirka commented:
Thank you both so much for the review!

> Removed cards disappear on device rotation, or when you navigate to the post and back, but I assume making them stay will overcomplicate things unnecessarily.

We can fix that, but as you say, it will complicate things a bit. We'd have to persist id's of removed posts and fetch one by one from the db after configuration change. 
AFAIK recommended approach is to show a snackbar with Undo action for couple of seconds. The Snackbar disappears after configuration change as well. Our solution is already way better as we can undo more than just last action. Wdyt @iamthomasbishop, is it worth it?


>I noticed that if you only have 1 saved post, then remove it, I don't see the same collapse transition/animation that I do when there are more than 1 posts. Example: https://cloudup.com/c-vBGusVxW3 . I also noticed that the loader/spinner is also appears oddly at the bottom-left of the screen.

I think when the list can fit on one screen (without scrolling), the last item is never animated. This is default Android behavior, so I'd suggest leave it as it is. Implementing our own animation just to fix this 'issue' seems like an overkill.

> When you collapse a card, if you look closely, you'll see that the contents are actually outside the bounds of the card as they fade out. Can we get the cards' contents to not overflow the bounds of the card while the transition is happening?

AFAIK Google has implemented it this way due to performance reasons. 
There are two states 
- full post
- removed post

The heigh (bounds) of the 'full post' view is actually never changed, the only thing the animation changes is 
- alpha of the 'full post' view 
- positions on the screen of all other views in the list. 
- alpha of the 'removed post' view

Changing bounds of a view results in measure/layout/draw cycle for each view in the list, which is quite costly.

So I think we can't actually fix that (without a performance impact) or I don't know how at least.

>When I mocked this up, I used the old lighter color icons, forgot we had darkened the colors of those during the recent accessibility audit. Can we make this save icon the darker color to match the other action icons?

There is a mess in the action icon colors :P
- Comments - there is just one state -> dark grey 
                      - on click -> light blue
- Like - disabled state -> dark grey
          - enabled state -> orange
         - on click -> light blue
- Save post - disabled -> light grey
                    - enabled -> light blue
                   - on click -> enabled light grey/disabled light blue
![icon_color](https://user-images.githubusercontent.com/2261188/40226376-d60c7f78-5a8b-11e8-91f7-77bb8b02a856.gif)

Can you please create a ticket for this and for the spacing issue so this PR doesn't get out of hands. I'll fix it ASAP. Thanks!!!

>I should've noted this earlier, but because we are removing the Share action from the top-level list of actions to make room for Save, I'd like to add Share (or "Share this post") to the ellipsis menu

No worries at all;). I just need some clarification. We've removed `comments` and `like` actions from the top-level list. By `Share action` you mean `like`? Or am I missing something?:) Thanks!

If you really mean the `like` action, we have to keep in mind that the state (liked/not liked) might be out of sync with server in the Saved Posts list. 

@khaykov I've done some refactoring while trying to fix the overflowing content issue. I'm sorry for adding more code for review, but I feel it's way cleaner now. Thanks!

iamthomasbishop commented:
Regarding the disappearing posts on orientation change, because the collapsed posts are acting as the undo/snackbar, we should make sure these don't disappear until the user leaves the screen. I would consider fixing this asap, but I don't know if it's a blocker for v1.

> This is default Android behavior, so I'd suggest leave it as it is. Implementing our own animation just to fix this 'issue' seems like an overkill.

Fair enough! Not the end of the world.

> I believe it's at the same place as on all other Reader's screens. But I can definitely move it, if you tell me where.

Let's get that thing centered :)

> There is a mess in the action icon colors :P

What I am most worried about is the inactive/unsaved style matching the darker grey of the others, so can we make sure that's matching? It actually doesn't bug me that the active colors of the star and saved icons are different colors, but that's something we can revisit.

> I just need some clarification. We've removed comments and like actions from the top-level list. By Share action you mean like?

Sorry, I was mixing up iOS + Android. On iOS we have a `Share` action on the post actions that we ended up hiding, but kept `Share` in the overflow menu :) In that case, the `Share` action would be a new item - so ignore me. I'd like to make iOS, Android, and Calypso consistent in terms of actions, but that's for a different discussion :)

malinajirka commented:
Thank you @iamthomasbishop !

I've fixed color of the bookmark button and position of the progress indicator.

![fixes](https://user-images.githubusercontent.com/2261188/40410617-8a7a3cb2-5e6e-11e8-9a2b-c9d50e0e06cb.png)

>because the collapsed posts are acting as the undo/snackbar, we should make sure these don't disappear until the user leaves the screen.

I've created a new github issue #7800, as this will require some more work and I'd like to merge `save-post-for-later` feature ASAP.
However, tbh I'm not sure I understand your reasoning. The undo/snackbar disappears on orientation change or after +-3seconds by default.

@khaykov It's ready for another round. Thanks!!! ;-)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderPostAdapter.java:
```diff
@@ -364,7 +394,32 @@ private void renderXPost(int position, ReaderXPostViewHolder holder) {
 
-    private void renderPost(int position, ReaderPostViewHolder holder) {
+    private void renderRemovedPost(final int position, final ReaderRemovedPostViewHolder holder) {
+        final ReaderPost post = getItem(position);
```
I believe it can't be null, unless there is a bug in the code. So I'd suggest sticking with Fail ASAP principal and we'll see:P. Wdyt?

malinajirka commented:
Thanks @khaykov!

I noticed the app was crashing on api 19 because of the vector drawableLeft. So I have fixed the crash along with the weird behavior of the undo icon.

Ready for another (hopefully final) round;). Thanks!!!
